src/main/java/com/hazelcast/kubernetes/KubernetesConfig.java

```diff
@@ -0,0 +1,297 @@
+import static com.hazelcast.kubernetes.KubernetesProperties.SERVICE_LABEL_VALUE;
+import static com.hazelcast.kubernetes.KubernetesProperties.SERVICE_NAME;
+import static com.hazelcast.kubernetes.KubernetesProperties.SERVICE_PORT;
+
+/**
+ * Responsible for fetching, parsing, and validating Hazelcast Kubernetes Discovery Strategy input properties.
+ */
+final class KubernetesConfig {
+    private static final String DEFAULT_MASTER_URL = "https://kubernetes.default.svc";
+    private static final int DEFAULT_SERVICE_DNS_TIMEOUT_SECONDS = 5;
+    private static final int DEFAULT_KUBERNETES_API_RETRIES = 3;
+
+    // Parameters for DNS Lookup mode
+    private final String serviceDns;
+    private final int serviceDnsTimeout;
+
+    // Parameters for Kubernetes API mode
+    private final String serviceName;
+    private final String serviceLabelName;
+    private final String serviceLabelValue;
+    private final String namespace;
+    private final boolean resolveNotReadyAddresses;
+    private final int kubernetesApiRetries;
+    private final String kubernetesMasterUrl;
+    private final String kubernetesApiToken;
+    private final String kubernetesCaCertificate;
+
+    // Parameters for both DNS Lookup and Kubernetes API modes
+    private final int servicePort;
+
+    KubernetesConfig(Map<String, Comparable> properties) {
+        this.serviceDns = getOrNull(properties, KUBERNETES_SYSTEM_PREFIX, SERVICE_DNS);
+        this.serviceDnsTimeout
+                = getOrDefault(properties, KUBERNETES_SYSTEM_PREFIX, SERVICE_DNS_TIMEOUT, DEFAULT_SERVICE_DNS_TIMEOUT_SECONDS);
+        this.serviceName = getOrNull(properties, KUBERNETES_SYSTEM_PREFIX, SERVICE_NAME);
+        this.serviceLabelName = getOrNull(properties, KUBERNETES_SYSTEM_PREFIX, SERVICE_LABEL_NAME);
+        this.serviceLabelValue = getOrDefault(properties, KUBERNETES_SYSTEM_PREFIX, SERVICE_LABEL_VALUE, "true");
+        this.namespace = getOrDefault(properties, KUBERNETES_SYSTEM_PREFIX, NAMESPACE, getNamespaceOrDefault());
+        this.resolveNotReadyAddresses = getOrDefault(properties, KUBERNETES_SYSTEM_PREFIX, RESOLVE_NOT_READY_ADDRESSES, false);
+        this.kubernetesApiRetries
+                = getOrDefault(properties, KUBERNETES_SYSTEM_PREFIX, KUBERNETES_API_RETIRES, DEFAULT_KUBERNETES_API_RETRIES);
+        this.kubernetesMasterUrl = getOrDefault(properties, KUBERNETES_SYSTEM_PREFIX, KUBERNETES_MASTER_URL, DEFAULT_MASTER_URL);
+        this.kubernetesApiToken = getApiToken(properties);
+        this.kubernetesCaCertificate = caCertificate(properties);
+        this.servicePort = getOrDefault(properties, KUBERNETES_SYSTEM_PREFIX, SERVICE_PORT, 0);
+
+        validateConfig();
+    }
+
+    private String getNamespaceOrDefault() {
+        String namespace = System.getenv("KUBERNETES_NAMESPACE");
+        if (namespace == null) {
+            namespace = System.getenv("OPENSHIFT_BUILD_NAMESPACE");
```

Comment:
Why do we have openshift specific customization here?

Reply:
I actually never used these variables, but looking at [this](https://docs.openshift.com/container-platform/3.4/dev_guide/builds/build_output.html) my guess is that the OpenShift variable is automatically set when you create an image stream with OpenShift tooling.